### PR TITLE
8286431: Do not use resource array in posix mmap_attach_shared()

### PR DESCRIPTION
**Pull request by iklam**
Please review this small clean up

perfMemory_posix.cpp uses a lot of explicit C heap alloc/free because the function may be called during early VM bootstrap, before ResourceMark can be used.

However, we have a single odd use of resource allocation in this file. It's supposed to work with Java exceptions. However, the code actually makes an explicitly exception check, so there's no need to do the resource alloc at all.

Also, there's no need to close the fd in case of exception.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286431](https://bugs.openjdk.org/browse/JDK-8286431): Do not use resource array in posix mmap_attach_shared()

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [6e99ca65](https://git.openjdk.org/jdk/pull/10686/files/6e99ca65518b50d3e2528713c428eca46715f37d)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer) ⚠️ Review applies to [6e99ca65](https://git.openjdk.org/jdk/pull/10686/files/6e99ca65518b50d3e2528713c428eca46715f37d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10686/head:pull/10686` \
`$ git checkout pull/10686`

Update a local copy of the PR: \
`$ git checkout pull/10686` \
`$ git pull https://git.openjdk.org/jdk pull/10686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10686`

View PR using the GUI difftool: \
`$ git pr show -t 10686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10686.diff">https://git.openjdk.org/jdk/pull/10686.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/posix/perfMemory_posix.cpp [hunk 1]
```diff
@@ -1196,29 +1194,19 @@ static void mmap_attach_shared(int vmid, char** addr, size_t* sizep, TRAPS) {
...
+  // open the shared memory file for the give vmid
   char* filename = get_sharedmem_filename(dirname, vmid, nspid);
 
-  // copy heap memory to resource memory. the open_sharedmem_file
-  // method below need to use the filename, but could throw an
-  // exception. using a resource array prevents the leak that
-  // would otherwise occur.
-  char* rfilename = NEW_RESOURCE_ARRAY(char, strlen(filename) + 1);
-  strcpy(rfilename, filename);
+  // Handle the exception explicitly so we can free the strings.
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/perfMemory_posix.cpp:1200]: Suggestion:
// We don't use CHECK as we need to free the strings even if an exception occurred.

File: src/hotspot/os/posix/perfMemory_posix.cpp [hunk 2]
```diff
@@ -1196,29 +1194,19 @@ static void mmap_attach_shared(int vmid, char** addr, size_t* sizep, TRAPS) {
...
   FREE_C_HEAP_ARRAY(char, luser);
   FREE_C_HEAP_ARRAY(char, dirname);
   FREE_C_HEAP_ARRAY(char, filename);
 
-  // open the shared memory file for the give vmid
-  int fd = open_sharedmem_file(rfilename, file_flags, THREAD);
-
-  if (fd == OS_ERR) {
-    return;
-  }
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/perfMemory_posix.cpp:1218]: I think you still need this for the case where the file is insecure - no exception is thrown in that case.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
One issue below and a suggestion.

I'm assuming we used the CHECK macro at one stage - hence the use of the RM etc.

File: src/hotspot/os/posix/perfMemory_posix.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/os/posix/perfMemory_posix.cpp]: Fixed.

File: src/hotspot/os/posix/perfMemory_posix.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/os/posix/perfMemory_posix.cpp:1218]: Fixed.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks

**Review by jdksjolen — APPROVED** [review]
This looks good to me, thanks for the changes.

**Comment by iklam** [comment]
Thanks @dholmes-ora and @jdksjolen for the review.
/integrate

**Pull request closed without merging**